### ISSUE TITLE
[Lite] Define the common/specialized boundary for model composition, checkpointing, and validation

### ISSUE THREAD

**Issue by ISEEKYAN** [ISEEKYAN/Megatron-LM#118]
## Core question

The central design problem in Megatron Lite is not simply how to remove duplication. It is how to draw a durable boundary between:

- invariants that should be implemented and verified once by the runtime; and
- model semantics that must remain easy to specialize without fighting a generic framework.

Both extremes are harmful:

- **Too little sharing** duplicates distributed lifecycle code in every model, so fixes land unevenly and behavior silently diverges.
- **Too much sharing** creates a lowest-common-denominator abstraction or “hook soup” that obscures control flow and makes genuinely new model behavior harder to implement.

The goal should be a small, typed, testable common composition kernel with explicit specialization points—not maximum deduplication.

## Why this needs attention now

DeepSeek-V4, GLM-5, Kimi-K2, Qwen3.5, and Qwen3-MoE currently repeat substantial protocol-level orchestration:

- parallel initialization and topology validation;
- pipeline chunk construction and layer assignment;
- model initialization and device placement;
- dist-opt/FSDP2/no-optimizer selection;
- delayed FSDP2 optimizer construction after HF loading;
- finalize-grad wiring;
- auxiliary-loss/pre-forward hooks;
- optimizer backend metadata;
- `post_model_load_hook` and `ModelBundle.extras` assembly;
- common checkpoint and forward-output glue.

At the same time, these models have real differences that should remain model-owned:

- attention and sequence-layout semantics;
- MoE/expert packing and placement;
- router state and replay;
- MTP and auxiliary heads;
- topology restrictions;
- persistent model buffers;
- HF/native tensor transforms and quantized resync formats;
- model-specific optimizer/wrapping exceptions.

The current boundary is implicit. As a result, code that looks shared is not necessarily authoritative, while model protocols contain both semantic specialization and duplicated runtime machinery.

## Concrete failure: HF checkpoint save

shyoshyo/Megatron-LM@2be71877ec71d6ab62fa08d0a948f0cecdae853f illustrates the problem:

- it correctly identifies that `verl_mlite` does not honor `checkpoint.save_contents=["hf_model"]`;
- it changes the generic HF saver to stream 5 GiB shards instead of materializing the full state dict;
- Qwen3.5 and Qwen3-MoE use that generic saver;
- DeepSeek-V4, GLM-5, and Kimi-K2 each have a model-local `save_hf_weights` that still executes `dict(export_hf_weights(..., cpu=True))`;
- the DS4 protocol calls its local saver directly, so the memory fix does not reach DS4 and rank 0 can still materialize the whole HF checkpoint.

This is not primarily a missing bridge function. It shows that the codebase has no enforced answer to two questions:

1. Which part of HF export is invariant infrastructure?
2. Which part is model-specific tensor semantics?

The tensor mapping and model-owned buffers are specialization. Transactional streaming, shard/index generation, staging, memory bounds, and failure behavior are common infrastructure.

### 1. Share invariants, not model semantics

The common layer should own behavior whose correctness contract is identical across models:

- lifecycle phase ordering;
- parallel-state initialization;
- standard optimizer state transitions;
- delayed post-load construction;
- `ModelBundle` assembly and required metadata;
- checkpoint transactions and serialization mechanics;
- distributed participation rules;
- error handling, cleanup, and telemetry;
- conformance validation.

The model layer should own semantic choices and tensor transformations.

### 2. Specialization points must be typed and phase-oriented

Avoid a monolithic `compose_model()` with dozens of optional callbacks. Define a small number of typed capabilities around lifecycle phases, for example:

```python
class ModelCompositionSpec(Protocol):
    def validate(self, model_cfg, impl_cfg) -> None: ...
    def build_chunks(self, model_cfg, impl_cfg, parallel_state) -> list[nn.Module]: ...
    def configure_chunks(self, chunks, context) -> None: ...
    def forward_step(self, model, batch): ...
    def checkpoint_adapter(self) -> CheckpointAdapter: ...
```

The shared composer can then own the invariant state machine:

```python
compose_model_bundle(spec, model_cfg, impl_cfg) -> ModelBundle
```

Model-specific overrides should be explicit capabilities, not incidental checks for attributes or model names.

### 3. Keep an explicit escape hatch

New architectures will occasionally need to replace a lifecycle phase. That should be possible without copying the entire composer.

An escape hatch should be:

- explicit in the model spec;
- scoped to one phase;
- visible in logs/metadata;
- covered by a model-specific test;
- prohibited from silently changing unrelated common semantics.

### 4. Make the common path authoritative

If a concern is classified as common infrastructure, model code should not be able to accidentally bypass it.

For HF saving, model adapters should yield a model-specific tensor stream, while one common writer owns serialization:

```python
write_hf_tensor_stream(
    model_checkpoint_adapter.export_hf_tensors(...),
    output_dir,
    max_shard_size=...,
)
```

The writer should provide:

- bounded-memory sharding;
- standard HF shard names and index metadata;
- duplicate-key detection;
- temporary-directory writes followed by atomic promotion;
- stale artifact cleanup;
- implemented local staging for non-mmap filesystems;
- deterministic ordering and failure cleanup;
- honest memory accounting.

Model code should only own names, tensor transforms, model-specific buffers, and export ordering constraints.

## veRL checkpoint integration

`MegatronLiteEngine` should call a supported runtime checkpoint capability rather than reaching into `ModelHandle._extras` and protocol internals.

The public contract should distinguish:

1. resumable distributed training checkpoint save/load;
2. HF export for deployment or cold loading;
3. optional config/tokenizer/processor materialization.

Additional requirements:

- `hf_model` remains opt-in, consistent with veRL's default `[model, optimizer, extra]`;
- missing `save_contents` must not trigger an unexpected full-model export;
- an explicit request either produces a valid artifact or fails loudly;
- partial failures must not leave a checkpoint that appears complete.

## Validation strategy: test both sides of the boundary

A shared abstraction is only useful if tests prove both its invariants and its specialization behavior.

### Common conformance suite

Run against every registered Lite model:

- build and `ModelBundle` contract;
- optimizer lifecycle for dist-opt, FSDP2, and no optimizer;
- delayed post-HF-load initialization;
- distributed save -> fresh build -> load;
- HF tensor stream -> sharded save -> cold load;
- offload -> reload -> save;
- PP/EP/CP participation and completeness;
- required model-owned persistent buffers;
- failure cleanup and actionable errors.

The suite should fail if a model silently replaces or bypasses an invariant common phase.

### Model-specific semantic tests

Keep focused tests for behavior that should not be generalized:

- attention/THD/CP semantics;
- expert packing and global/local expert mapping;
- router state and replay;
- MTP behavior;
- custom buffers and auxiliary heads;
- quantized resync formats;
- model-specific topology restrictions.

### Checkpoint writer tests

- single and multiple shards;
- duplicate keys;
- injected write/rename failure and cleanup;
- overwrite multi-shard -> single-shard without stale index data;
- deterministic weight map;
- bounded peak host memory;
- long writer stalls versus distributed collective timeout behavior.

### GPU tiers

- CPU/unit tests for contracts and serialization;
- 2/8-GPU proxy tests for PP/EP/CP and optimizer backends;
- full-size delivery tests only after proxy conformance passes;
- record rank-0 RSS, GPU peak, elapsed time, and artifact fingerprints.

## Acceptance criteria

- [ ] The project documents which lifecycle concerns are common invariants and which are model-owned specialization.
- [ ] A typed composition kernel owns the standard lifecycle without becoming an unstructured callback registry.
- [ ] All five current Lite model protocols use the common lifecycle for invariant phases.
- [ ] Each model's remaining compose code primarily expresses semantic specialization and explicit exceptions.
- [ ] Phase-scoped escape hatches are visible and tested.
- [ ] All model HF adapters feed one transactional streaming writer.
- [ ] DS4, GLM-5, and Kimi-K2 no longer materialize `dict(export_hf_weights(...))` during HF save.
- [ ] `verl_mlite` supports explicit `hf_model` export without changing default checkpoint behavior.
- [ ] Rewriting a checkpoint cannot load stale shard/index data.
- [ ] Cross-model conformance tests and model-specific semantic tests are both required for new integrations.
- [ ] Full-size validation demonstrates bounded rank-0 memory.

## Non-goals

- Forcing different model architectures into identical forward logic.
- Hiding model-specific topology or optimizer constraints.
- Replacing clear model code with generic indirection solely to reduce line count.
- Treating every hook as evidence that it belongs in the common layer.

## References

- Motivating HF save change: https://github.com/shyoshyo/Megatron-LM/commit/2be71877ec71d6ab62fa08d0a948f0cecdae853f
- veRL default checkpoint contents: https://github.com/verl-project/verl/blob/main/verl/trainer/config/_generated_ppo_trainer.yaml
- Relevant model protocols:
  - `experimental/lite/megatron/lite/model/deepseek_v4/lite/protocol.py`
  - `experimental/lite/megatron/lite/model/glm5/lite/protocol.py`
  - `experimental/lite/megatron/lite/model/kimi_k2/lite/protocol.py`
  - `experimental/lite/megatron/lite/model/qwen3_5/lite/protocol.py`
  - `experimental/lite/megatron/lite/model/qwen3_moe/lite/protocol.py`
